pkg/storegateway/indexheader/reader_pool_test.go

```
@@ -7,104 +7,145 @@ package indexheader
+			lazyReaderIdleTimeout:  time.Minute,
+			eagerLoadReaderEnabled: true,
+			initialSync:            true,
+			expectedLoadCountMetricBeforeLabelNamesCall: 1, // the index header will be eagerly loaded before the operation
+			expectedLoadCountMetricAfterLabelNamesCall:  1,
+			createLazyLoadedHeadersSnapshotFn: func(blockId ulid.ULID) lazyLoadedHeadersSnapshot {
+				return lazyLoadedHeadersSnapshot{
+					IndexHeaderLastUsedTime: map[ulid.ULID]int64{blockId: time.Now().UnixMilli()},
+					UserID:                  "anonymous",
+				}
+			},
+		},
+		"block is present in pre-shutdown loaded blocks and eager-loading is enabled, loading index header after initial sync": {
+			lazyReaderEnabled:      true,
+			lazyReaderIdleTimeout:  time.Minute,
+			eagerLoadReaderEnabled: true,
+			initialSync:            false,
+			expectedLoadCountMetricBeforeLabelNamesCall: 0, // the index header is not eager loaded if not during initial-sync
+			expectedLoadCountMetricAfterLabelNamesCall:  1,
+			createLazyLoadedHeadersSnapshotFn: func(blockId ulid.ULID) lazyLoadedHeadersSnapshot {
+				return lazyLoadedHeadersSnapshot{
+					IndexHeaderLastUsedTime: map[ulid.ULID]int64{blockId: time.Now().UnixMilli()},
+					UserID:                  "anonymous",
+				}
+			},
+		},
+		"block is not present in pre-shutdown loaded blocks snapshot and eager-loading is enabled": {
+			lazyReaderEnabled:      true,
+			lazyReaderIdleTimeout:  time.Minute,
+			eagerLoadReaderEnabled: true,
+			initialSync:            true,
+			expectedLoadCountMetricBeforeLabelNamesCall: 0, // although eager loading is enabled, this test will not do eager loading because the block ID is not in the lazy loaded file.
+			expectedLoadCountMetricAfterLabelNamesCall:  1,
+			createLazyLoadedHeadersSnapshotFn: func(_ ulid.ULID) lazyLoadedHeadersSnapshot {
+				// let's create a random fake blockID to be stored in lazy loaded headers file
+				fakeBlockID, _ := ulid.New(ulid.Now(), rand.Reader)
```

nit: you can use ulid.MustNew to make sure the test doesn't silently fail